pages/docs/v3/core-apis/data-types.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Capacitor data types
+description: Data types in Capacitor
+---
+
+# Capacitor data types
+
+Data moving between the web runtime and native environments in Capacitor have to be serialized and deserialized so that they can be stored natively in each language. The supported data types are those that can be represented in JSON such as numbers, strings, booleans, arrays, and objects (or dictionaries or key-value stores).
+
+## iOS
+
+While Swift is the preferred language on iOS, it interoperates with Objective-C (upon which the system frameworks are built) and so the platform supports the intersection of three languages. Most data types will be translated as expected but there are some cases that may require special attention.
+
+### Null Values
+
+Objective-C does not support storing null values in collections such as arrays, dictionaries, or sets. Instead it uses a special placeholder object, [`NSNull`](https://developer.apple.com/documentation/foundation/nsnull?language=objc), to represent a null value. In contrast, Swift uses [Optionals](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html) to describe a value that might be null. Swift can manipulate `NSNull` values but Objective-C cannot handle Optionals (although, in some contexts, the runtime will automatically map optionals into the underlying value or `NSNull`). These `NSNull` objects can appear regardless of which language you are using.
+
+As an example, consider the following object being passed to a Capacitor plugin call:
+
+```typescript
+{ 'foo': null, 'bar': [1, 2, null, 4]}
+```
+
+#### Dictionaries
+
+`CAPPluginCall` stores this data as its `options` dictionary property but has a variety of convenience accessors that operate on it. The convenience accessors will try to cast the value to the expected type(s) so `NSNull` values will get filtered out.
+
+```swift
+if let value = call.getString("foo") {
+    // false, `value` is nil
+}
+```
+
+However, accessing the storage property directly can return an `NSNull` object so its type will need to be checked before it is used.
+
+```swift
+let value = call.options["foo"]
+if let value = value {
+    // true, 'value' is a valid object
+}
+if let value = value as? Int {
+    // false, the NSNull object cannot be cast to Int
+}
+if let value = value as? NSNull {
+    // true, 'value' is a NSNull
+}
+```
+
+In general, it is not recommended to rely on the presence of a key to convey meaning. Only the corresponding value should be evaluated after its expected type is confirmed. Since dictionaries typically require the typing of each value as it is extracted, the inclusion of `NSNull` can mostly be ignored.
+
+#### Arrays
+
+Since accessing arrays typically requiring typing the whole collection, it is more important to consider if they are heterogeneous or not.
```

Comment:
1. 
```suggestion
Since accessing arrays typically require typing the whole collection, it is more important to consider if they are heterogeneous or not.
```

2. I would also consider defining "heterogeneous", or using simpler language.

Reply:
Fixed. I've tried for a contextual definition of heterogeneous.